Add ESLint to Project

nicholaschuayunzhi opened the pull request:
Adds ESLint and its lodash plugin as a dev dependency. --fix command not executed yet. The eslintrc file was taken from https://github.com/MarkBind/markbind/blob/master/.eslintrc.js

nicholaschuayunzhi committed:
Add eslint file

nicholaschuayunzhi committed:
Add eslintignore for minified

Gisonrg commented:
Nice one! Can't wait to fix all the style issues in the project :D

Before I merge this, do you think we should use an existing style standard for our eslint?
I am thinking about using either https://github.com/airbnb/javascript or https://github.com/google/eslint-config-google

Feel free to discuss with the team to choose a standard you all agreed on.

nicholaschuayunzhi commented:
I've used airbnb before. I would prefer airbnb as their docs have the eslint rules which we can ctrl f easily to see an example of.

damithc commented:
The common standards for our projects (open for updates) https://github.com/oss-generic/process/blob/master/docs/CodingStandards.adoc

Gisonrg commented:
Let's use the airbnb preset and modified based on our standard then :P

 @nicholaschuayunzhi Can you help us set up the eslint as discussed?

acjh commented:
> The common standards for our projects (open for updates) https:\//github.com/oss-generic/process/blob/master/docs/CodingStandards.adoc
>> If a topic is not covered in this document, refer to airbnb JavaScript Style Guide

airbnb it is! Also, [teammates-eslint.yml](https://github.com/TEAMMATES/teammates/blob/60a9efd3b5f1c7e1e0d035160bf9ed1f53e5d800/static-analysis/teammates-eslint.yml).

nicholaschuayunzhi commented:
Okay I'll work on it. Need to do some research first.

Also do you want it to be standardized across all the modules?

Gisonrg commented:
Nice, take you time. Let's start from `markbind-cli` first, then extend it to `markbind` (vue modules has different style standard).

nicholaschuayunzhi commented:
Changed rules wrt to [oss-generic coding standards](https://docs.google.com/document/d/1gZ6WG6HBTJYHAtVkz9kzi_SUuzfXqzO-SvFnLuag2xM/pub?embedded=true)

Rules that can't be modified with existing eslint rules:
- Constants named in screaming snake case
- Prefer higher-level breaks to lower-level breaks. In the example below, the first is preferred, since the break occurs outside the parenthesized expression, which is at a higher level.
- suggestion for ternary expression
![image](https://user-images.githubusercontent.com/22221132/35140362-e84fb3b0-fd31-11e7-8ea6-7037525153cb.png)

Read up on writing custom rules. Is it worth the time? Anyone with experience? Also will add to dev guide on how to set up eslint.

nicholaschuayunzhi commented:
I've updated the eslint config as discussed. Disabled func-names rule and added line break style.

acjh commented on .eslintrc.js:
```diff
@@ -0,0 +1,25 @@
+  "extends": "airbnb-base",
+  "rules": {
+    "function-paren-newline": "off",
+    "func-names": "off",
```
Alphabetical (ascii) order:
- `func-names` before `function-paren-newline`
- `FunctionDeclaration` before `FunctionExpression`

acjh commented on .eslintrc.js:
```diff
@@ -0,0 +1,25 @@
+        "CallExpression": { "arguments": "first" } ,
+        "FunctionExpression": { "parameters": "first" },
+        "FunctionDeclaration": { "parameters": "first" },
+        "MemberExpression": 1,
```
`1` seems to be the default in [indent](https://eslint.org/docs/rules/indent.html) and is not overridden by [airbnb](https://github.com/airbnb/javascript).

acjh commented:
I took a while to understand why some choices were made.

Perhaps we should document this in the Wiki, together with the setup instructions.

acjh commented on .eslintrc.js:
```diff
@@ -0,0 +1,24 @@
+        "ArrayExpression": "first",
+        "CallExpression": { "arguments": "first" } ,
+        "FunctionExpression": { "parameters": "first" },
+        "FunctionDeclaration": { "parameters": "first" },
```
Alphabetical (ascii) order:
- `FunctionDeclaration` before `FunctionExpression`

acjh commented:
> Perhaps we should document this in the Wiki, together with the setup instructions.

Maybe *not* with the setup instructions (seems out-of-place in the [Developer Guide](https://github.com/MarkBind/markbind-cli/wiki/Developer-Guide)).

nicholaschuayunzhi commented:
I'm working on the explanation in a separate gist i can link to in the dev guide. Sorry for missing out the previous ordering of keys.

nicholaschuayunzhi commented:
I've removed ArrayExpression option in indent as the default option (1) account for aligning elements based on 2 spaces of indentation.

I've added the `"array-bracket-newline": ["error", { "multiline": true }],` to be consistent with airbnb's
`object-curly-newline` rule, as well as to prevent such code from passing which is not wanted based on nus oss js standard.
```
let array = [1,
  2,
  3,]
```

nicholaschuayunzhi commented:
I've added a [gist](https://gist.github.com/nicholaschuayunzhi/bfe53dbb5f1a0f02d545d55956f9ab7e) for the explanation to the developer guide.

acjh commented:
> to be consistent with airbnb's `object-curly-newline` rule

I don't see this 🤔

acjh commented on .eslintrc.js:
```diff
@@ -0,0 +1,24 @@
+      "error",
+      2,
+      {
+        "CallExpression": { "arguments": "first" } ,
```
Remove space before comma.

acjh commented:
> I've added a gist for the explanation to the developer guide.

Nice. This can be a page in the Wiki. Let's [create a sidebar](https://help.github.com/articles/creating-a-sidebar/); the current pages need categorisation anyway.

nicholaschuayunzhi commented:
> to be consistent with airbnb's object-curly-newline rule
>>I don't see this 🤔

Oh dear, i think i saw wrongly. But the rationale still stands to prevent the below from happening.
```
let array = [1,
  2,
  3,]
```

Weird though, my linter is firing off that error. Could it be a default rule? https://eslint.org/docs/rules/object-curly-newline

acjh commented:
Both rules seem to be `{ "multiline": true }` by default.

nicholaschuayunzhi commented:
Seems when i do not specify array-bracket-newline, it does not fire off any linting error.
![image](https://user-images.githubusercontent.com/22221132/35190814-e41128ea-fea5-11e7-856b-5814eb2ca9d0.png)

Need to look into it further.

nicholaschuayunzhi commented:
In the eslint docs for both rules, seems like [array-bracket-newline](https://eslint.org/docs/rules/array-bracket-newline) is not enabled by default.  
![image](https://user-images.githubusercontent.com/22221132/35190842-cf8fac1a-fea6-11e7-8049-55dc81ba1b7a.png)



However for [object-curly-newline](https://eslint.org/docs/rules/object-curly-newline). It is enabled.
![image](https://user-images.githubusercontent.com/22221132/35190840-a7e8302e-fea6-11e7-842b-134d5913f12c.png)

So i think we should still have the `array-bracket-newline` rule.

acjh commented:
Nice catch.

nicholaschuayunzhi commented:
I still find it weird though, according to [user guide](https://eslint.org/docs/rules/), no rules are enabled by default, but the docs seem to say otherwise.

nicholaschuayunzhi commented:
Ok i think i got it, they disabled `array-bracket-newline` and enabled `object-curly-newline` in airbnb-base preset. And it is not documented on the readme.

https://github.com/airbnb/javascript/blob/master/packages/eslint-config-airbnb-base/rules/style.js

nicholaschuayunzhi commented:
Okay I think it should be ready for review. Note that operator-linebreak is manually enabled as [airbnb-base package with enabled operator-linebreak rule is not published yet](https://github.com/airbnb/javascript/commit/6e1f11185d95432ba90a647efc027bd32983f17d)

I've also removed the line endings rule as airbnb-base already checks for unix endings

Gisonrg commented:
LGTM, feel free to merge this @acjh

acjh commented:
@nicholaschuayunzhi Thanks for your contribution!

acjh commented:
@nicholaschuayunzhi Can you create a PR to apply `--fix`?

nicholaschuayunzhi commented:
@acjh  Will do!
